_posts/to-api-or-to-not-api.md

```diff
@@ -0,0 +1,64 @@
+---
+layout: post
+header_image: /img/posts/to-api-or-to-not-api/header.jpg
+title: "To API, or not to API. Is this a real problem?"
+author: Mauro Servienti
+synopsis: "Breaking a public API is scary but sometimes unavoidable. Versioning an API is complex and error-prone. In many cases, I observed teams breaking an API that should not have been there to begin with, or fighting with versioning issues. The problem, though, was not the API."
+tags:
+- Design
+- Architecture
+- SOA
+- Coupling
+- Versioning
+---
+
+I'm taking this opportunity to answer another question that got asked during one of my last presentations:
+
+> How to design evolvable APIs in a distributed system?
+
+Before being able to answer that question, we need to dive a bit into the problem space and ask a different question first:
+
+> Where do we need APIs?
+
+One could argue that we need APIs in three places:
+
+1. To serve content to the UI (or to clients consuming data, UI is one of these)
+2. To integrate with third parties
+3. To allow cross-service communication 
+
+Let's start from the end. In ["Slice it"](https://milestone.topics.it/view-model-composition/2019/04/09/slice-it.html), I introduced vertical slices and their relationship with the ViewModel Composition architectural concept. In my ["All our aggregates are wrong"](https://milestone.topics.it/talks/all-our-aggregates-are-wrong.html) talk, I look at vertical slices from a slightly different angle. Towards the end, I state that once we have full vertical slices, thanks to services isolation, cross-service communication can be implemented using thin events that can be as thin as name and identifier.
+
+![slide screenshot](/img/posts/to-api-or-to-not-api/screenshot.png){:class="img-fluid mx-auto d-block"}
+
+> The presentation recording is available on YouTube; for example, [this one from the NDC Conference](https://youtu.be/hev65ozmYPI) in Copenhagen.
+
+The essence is that there is no cross-service communication of any sort that requires sharing data. That means we don't need APIs to allow services to invoke other services. If the owning service already stores, there is no need for a service to request any data from other services, or microservice if you will. If data are misplaced, then we have a service boundaries problem that we need to address.
```

Comment:
Yes, "data" is technically plural. But the debate has ended and consistency lost.

```suggestion
The essence is that there is no cross-service communication of any sort that requires sharing data. That means we don't need APIs to allow services to invoke other services. If the owning service already stores, there is no need for a service to request any data from other services, or microservice if you will. If data is misplaced, then we have a service boundaries problem that we need to address.
```